internal/plugincommon/discovery/processor.go

```diff
@@ -0,0 +1,120 @@
+package discovery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+
+	"github.com/smartcontractkit/chainlink-ccip/internal/plugincommon"
+	dt "github.com/smartcontractkit/chainlink-ccip/internal/plugincommon/discovery/discoverytypes"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/consts"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/reader"
+)
+
+// Ensure that PluginProcessor is implemented.
+var _ plugincommon.PluginProcessor[dt.Query, dt.Observation, dt.Outcome] = &ContractDiscoveryProcessor{}
+
+// ContractDiscoveryProcessor is a plugin processor for discovering contracts.
+type ContractDiscoveryProcessor struct {
+	lggr      logger.Logger
+	reader    *reader.CCIPReader
```

Comment:
Why a pointer to an interface?

Reply:
We need to make sure that when the reader is updated here it is also updated in other parts of the plugin.

It works without the pointer because the shallow copies mean each slice refers to the same array, but will that always be the case?

Reply:
The implementation is a pointer receiver so it should work without the pointer. However, I kind of see what you're saying about the shared aspect of it.

The way the reader is used right now, its basically a singleton. The way things are designed right now, I don't see this being the case. Feels like a change in this would cause a super subtle bug though, gotta think

Reply:
Correct me if I'm wrong on this but I think the type assertion on the `ccipChainReader` type helps us avoid this problem.

```golang
// Interface compliance check - on the pointer type
var _ CCIPReader = (*ccipChainReader)(nil)
// Interface compliance check - on the value type - FAILS
var _ CCIPReader = ccipChainReader{}
```

So as long as:

* CCIP reader is a singleton
* interface compliance is on the pointer receiver

It should be safe to have this as just the interface rather than pointer to interface. We can put a comment on the compliance check to ensure that this cannot change and what the implications are.

Reply:
Maybe there's a test we can write to ensure that the pointer type implements the interface - not sure though, seems out of scope of the Go testing facilities since its part of the type system.

Reply:
There are examples in the stdlib of tricks to avoid accidentally copying things that shouldn't be copied.

strings.Builder
```
// A Builder is used to efficiently build a string using [Builder.Write] methods.
// It minimizes memory copying. The zero value is ready to use.
// Do not copy a non-zero Builder.
type Builder struct {
	addr *Builder // of receiver, to detect copies by value
	buf  []byte
}
```

io.File
```
// File represents an open file descriptor.
type File struct {
	*file // os specific
}
```

Maybe we should adopt one of these

Reply:
Or maybe leverage `sync.Mutex` not being copyable:

```golang
type nonCopyable struct {
    _ sync.Mutex
}

type ccipChainReader struct {
  nonCopyable
// .. rest
}
```